File: Introduction to Robotics Course/Chapter15/Lesson5/Chapter15_Lesson5.cpp
#include <vector>
#include <stdexcept>
#include <iostream>

struct LogEntry {
    int group;     // 0 or 1
    double cost;   // nonnegative cost for this interaction
};

double groupRisk(const std::vector<LogEntry> &log, int g) {
    double sum = 0.0;
    std::size_t count = 0;
    for (const auto &e : log) {
        if (e.group == g) {
            sum += e.cost;
            ++count;
        }
    }
    if (count == 0) {
        throw std::runtime_error("No samples for requested group");
    }
    return sum / static_cast<double>(count);
}

int main() {
    std::vector<LogEntry> log = {
        {0, 0.0}, {0, 1.0}, {1, 0.2}, {1, 0.1}, {1, 0.7}, {0, 0.3}
    };
    double r0 = groupRisk(log, 0);
    double r1 = groupRisk(log, 1);
    double disparity = std::abs(r1 - r0);

    std::cout << "r0 = " << r0 << ", r1 = " << r1
              << ", disparity = " << disparity << std::endl;
    const double eps = 0.05;
    if (disparity > eps) {
        std::cout << "Warning: fairness constraint violated." << std::endl;
    }
    return 0;
}
      

File: Introduction_to_Robotics_Course/Chapter15/Lesson5/Chapter15_Lesson5.py
import numpy as np

def group_risk(costs, groups, g):
    """
    costs : 1D numpy array of nonnegative costs for each interaction
    groups: 1D numpy array of group labels (0 or 1)
    g     : group of interest (0 or 1)
    """
    mask = (groups == g)
    if not np.any(mask):
        raise ValueError(f"No samples for group {g}")
    return float(costs[mask].mean())

def disparity(costs, groups):
    r0 = group_risk(costs, groups, 0)
    r1 = group_risk(costs, groups, 1)
    return abs(r1 - r0)

# Example usage with a simple threshold
eps = 0.05  # allowed disparity
costs = np.array([0.0, 1.0, 0.2, 0.1, 0.7, 0.3])
groups = np.array([0, 0, 1, 1, 1, 0])

d = disparity(costs, groups)
print("Estimated disparity:", d)
if d > eps:
    print("Warning: fairness constraint violated.")
      
